Math/NumberTheory/ArithmeticFunctions/Inverse.hs

```diff
@@ -0,0 +1,354 @@
+    divs = case divisorsListA of
+      ArithmeticFunction f g -> g $ mconcat $ map (uncurry f) fs
+
+    n :: a
+    n = product $ map (\(p, k) -> unPrime p ^ k) fs
+
+    pksSmall :: Map (Prime a) (S.Set Word)
+    pksSmall
+      = M.fromDistinctAscList
+      $ filter (not . null . snd)
+      $ map (\p -> (p, doPrime p))
+      $ takeWhile ((< lim) . unPrime)
+      $ primes
+
+    doPrime :: Prime a -> S.Set Word
+    doPrime (unPrime -> p) = S.fromDistinctAscList
+      [ e
+      | e <- [1 .. intToWord (integerLogBase (toInteger p) (toInteger n))]
+      , n `rem` ((p ^ (e + 1) - 1) `quot` (p - 1)) == 0
+      ]
+
+    pksLarge :: M.Map (Prime a) (S.Set Word)
+    pksLarge = M.unionsWith (<>)
+      [ maybe mempty (flip M.singleton (S.singleton e)) (isPrime p)
+      | d <- divs
+      , e <- [1 .. intToWord (integerLogBase (toInteger lim) (toInteger d))]
+      , let p = integerRoot e (d - 1)
+      , p ^ (e + 1) - 1 == d * (p - 1)
+      ]
+
+-- | Instead of multiplying all atomic series and filtering out everything,
+-- which is not divisible by @n@, we'd rather split all atomic series into
+-- a couple of batches, each of which corresponds to a prime factor of @n@.
+-- This allows us to crop resulting Dirichlet series (see 'filter' calls
+-- in 'invertFunction' below) at the end of each batch, saving time and memory.
+strategy
+  :: forall a c. (Euclidean c, Ord c)
+  => ArithmeticFunction a c
+  -- ^ Arithmetic function, which we aim to inverse
+  -> [(Prime c, Word)]
+  -- ^ Factorisation of a value of the arithmetic function
+  -> [PrimePowers a]
+  -- ^ Possible prime factors of an argument of the arithmetic function
+  -> [(Maybe (Prime c, Word), [PrimePowers a])]
+  -- ^ Batches, corresponding to each element of the factorisation of a value
+strategy tot fs tots = (Nothing, ret) : rets
+  where
+    fs' = sortBy (\(p1, _) (p2, _) -> p2 `compare` p1) fs
+
+    (ret, rets) = mapAccumL go tots fs'
+
+    go
+      :: [PrimePowers a]
+      -> (Prime c, Word)
+      -> ([PrimePowers a], (Maybe (Prime c, Word), [PrimePowers a]))
+    go ts (p, k) = (rs, (Just (p, k), qs))
+      where
+        predicate (PrimePowers q ls) = any (\l -> runTot q l `rem` unPrime p == 0) ls
+        (qs, rs) = partition predicate ts
+
+    runTot :: Prime a -> Word -> c
+    runTot p k = case tot of
+      ArithmeticFunction f g -> g $ f p k
+
+-- | Main workhorse.
+invertFunction
+  :: forall a b c.
+     (Num a, Ord a, Semiring b, Euclidean c, UniqueFactorisation c, Ord c)
+  => (a -> b)
+  -- ^ How to inject a number into a semiring
+  -> ArithmeticFunction a c
+  -- ^ Arithmetic function, which we aim to inverse
+  -> ([(Prime c, Word)] -> [PrimePowers a])
+  -- ^ How to find possible prime factors of the argument
+  -> c
+  -- ^ Value of the arithmetic function, which we aim to inverse
+  -> b
+  -- ^ Preimages
+invertFunction point f invF n
+  = DS.lookup n
+  $ foldl (\ds b -> uncurry processBatch b ds) (DS.fromDistinctAscList []) batches
```

Comment:
Not sure if it's worth it as an isolated change (meaning it's probably worth doing as part of a performance fixup PR, just fixing this line won't make much of a difference overall), but we should probably make an effort to avoid `foldl` and always rely on `foldl'`.